Python/optimizer_analysis.c

```diff
@@ -152,231 +157,136 @@ typedef struct _Py_UOpsAbstractInterpContext {
+    ctx->limit -= co_const_len;
+    if (ctx->limit <= ctx->water_level) {
         return NULL;
     }
     for (Py_ssize_t i = 0; i < co_const_len; i++) {
-        _Py_UOpsSymType *res = sym_init_const(ctx, PyTuple_GET_ITEM(co_consts, i), (int)i);
+        _Py_UOpsSymType *res = sym_init_const(ctx, PyTuple_GET_ITEM(co_consts, i));
         if (res == NULL) {
-            goto error;
+            return NULL;
         }
         sym_consts[i] = res;
     }
 
     return sym_consts;
-error:
-    PyMem_Free(sym_consts);
-    return NULL;
 }
 
+static inline _Py_UOpsSymType* sym_init_unknown(_Py_UOpsAbstractInterpContext *ctx);
 
-static inline _Py_UOpsSymType*
-sym_init_unknown(_Py_UOpsAbstractInterpContext *ctx);
+// 0 on success, anything else is error.
+static int
+ctx_frame_push(
+    _Py_UOpsAbstractInterpContext *ctx,
+    PyCodeObject *co,
+    _Py_UOpsSymType **localsplus_start,
+    int curr_stackentries
+)
+{
+    _Py_UOpsSymType **sym_consts = create_sym_consts(ctx, co->co_consts);
+    if (sym_consts == NULL) {
+        return -1;
+    }
+    assert(ctx->curr_frame_depth < MAX_ABSTRACT_FRAME_DEPTH);
+    _Py_UOpsAbstractFrame *frame = &ctx->frames[ctx->curr_frame_depth];
+    ctx->curr_frame_depth++;
 
-static void
-sym_copy_immutable_type_info(_Py_UOpsSymType *from_sym, _Py_UOpsSymType *to_sym);
+    frame->sym_consts = sym_consts;
+    frame->sym_consts_len = (int)Py_SIZE(co->co_consts);
+    frame->stack_len = co->co_stacksize;
+    frame->locals_len = co->co_nlocalsplus;
 
-/*
- * The reason why we have a separate frame_push and frame_initialize is to mimic
- * what CPython's frame push does. This also prepares for inlining.
- * */
-static inline int
-frame_push(_Py_UOpsAbstractInterpContext *ctx,
-           _Py_UOpsAbstractFrame *frame,
-           _Py_UOpsSymType **localsplus_start,
-           int locals_len,
-           int curr_stacklen,
-           int total_len)
-{
     frame->locals = localsplus_start;
-    frame->stack = frame->locals + locals_len;
-    frame->stack_pointer = frame->stack + curr_stacklen;
-    ctx->water_level = localsplus_start + total_len;
-    if (ctx->water_level > ctx->limit) {
+    frame->stack = frame->locals + co->co_nlocalsplus;
+    frame->stack_pointer = frame->stack + curr_stackentries;
+    ctx->water_level = localsplus_start + (co->co_nlocalsplus + co->co_stacksize);
+    if (ctx->water_level >= ctx->limit) {
         return -1;
     }
-    return 0;
-}
 
-static inline int
-frame_initalize(_Py_UOpsAbstractInterpContext *ctx, _Py_UOpsAbstractFrame *frame,
-                int locals_len, int curr_stacklen)
-{
+
     // Initialize with the initial state of all local variables
-    for (int i = 0; i < locals_len; i++) {
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
         _Py_UOpsSymType *local = sym_init_unknown(ctx);
         if (local == NULL) {
-            goto error;
+            return -1;
         }
         frame->locals[i] = local;
     }
```

Comment:
Isn't this wasteful when the first N locals are filled from the arguments? (In `_PUSH_FRAME`)